cpp/src/arrow/compare.cc

```diff
@@ -717,6 +719,83 @@ class TypeEqualsVisitor {
   bool result_;
 };
 
+#define VISIT_SIMPLE(SCALAR_TYPE)                                 \
+  Status Visit(const SCALAR_TYPE& left) {                         \
+    const auto& right = checked_cast<const SCALAR_TYPE&>(right_); \
+    result_ = right.value == left.value;                          \
+    return Status::OK();                                          \
+  }
```

Comment:
I wasn't too happy about this macro so I'm adding a common base class for the types that hit this case

Reply:
good for me!